### PR TITLE
Optimize performance

### PR DESCRIPTION
**Pull request by kelunik**
Ensure is_int and is_string avoid the namespace lookup and can use built-in optimizations.

Cache formatted hosts, as these are likely always the same in a server scenario.

Move conditions for formatting into caller to avoid additional function calls where possible.

### CODE REVIEW

**Comment by kelunik** [comment]
image

Before / after comparison for this example:

```php
#!/usr/bin/env php
<?php

use League\Uri\Http;

require dirname(__DIR__) . "/vendor/autoload.php";

for ($i = 0; $i < 1000000; $i++) {
    $uri = Http::createFromComponents([
        "scheme" => 'http',
        "host" => 'localhost',
        "port" => 1337,
        "path" => '/',
        "query" => '',
    ]);
}
```

**Comment by nyamsprod** [comment]
thanks for the optimizations 👍🏾

File: src/Uri.php [hunk 1]
```diff
@@ -221,7 +225,7 @@ private function __construct(
      */
     private function formatScheme(?string $scheme): ?string
     {
-        if (null === $scheme) {
+        if (null === $scheme || in_array($scheme, ['http', 'https', 'ws', 'wss'], true)) {
```
Comment (nyamsprod) [inline src/Uri.php:228]: after merging I have changed this line to 
```
        if (null === $scheme || isset(self::SCHEME_DEFAULT_PORT[$scheme]))
```

**Comment by nyamsprod** [comment]
after merging I have changed the line

> if (null === $scheme || in_array($scheme, ['http', 'https', 'ws', 'wss'], true))

to 

> if (null === $scheme || isset(self::SCHEME_DEFAULT_PORT[$scheme]))

it re-used a declared const and uses `isset` instead of `in_array` this should make it faster I guess.

Also I moved 

> static $formatRegisteredNameCache = [];
to be a static variable inside `formatHost` method.

File: src/Uri.php (hunk 1 above)
Comment (kelunik) [reply src/Uri.php:228]: 👍

**Pull request merged**